File: server/hotrod/src/main/scala/org/infinispan/server/hotrod/HotRodDecoder.scala
package org.infinispan.server.hotrod

import java.io.{IOException, StreamCorruptedException}
import java.lang.StringBuilder
import java.security.PrivilegedExceptionAction
import javax.security.auth.Subject
import javax.security.sasl.SaslServer

import io.netty.buffer.{ByteBuf, Unpooled}
import io.netty.channel._
import io.netty.handler.codec.ReplayingDecoder
import io.netty.util.CharsetUtil
import org.infinispan.manager.EmbeddedCacheManager
import org.infinispan.security.Security
import HotRodDecoderState._
import org.infinispan.server.core.Operation._
import org.infinispan.server.core._
import org.infinispan.server.core.logging.Log
import org.infinispan.server.core.security.AuthorizingCallbackHandler
import org.infinispan.server.core.transport.ExtendedByteBuf._
import org.infinispan.server.core.transport._

/**
 * Top level Hot Rod decoder that after figuring out the version, delegates the rest of the reading to the
 * corresponding versioned decoder.
 *
 * @author Galder Zamarreño
 * @author gustavonalle
 * @since 4.1
 */
class HotRodDecoder(cacheManager: EmbeddedCacheManager, val transport: NettyTransport, server: HotRodServer)
extends ReplayingDecoder[HotRodDecoderState](DECODE_HEADER) with StatsChannelHandler with ServerConstants with Constants with Log {

   val secure = server.getConfiguration.authentication().enabled()

   private val decodeCtx = new CacheDecodeContext(server)

   var saslServer: SaslServer = null
   var callbackHandler: AuthorizingCallbackHandler = null
   var subject: Subject = ANONYMOUS

   def decode(ctx: ChannelHandlerContext, in: ByteBuf, out: java.util.List[AnyRef]): Unit = {
      try {
         if (decodeCtx.isTrace) trace("Decode using instance @%x", System.identityHashCode(this))
         wrapSecurity {
            state match {
               case DECODE_HEADER => decodeHeader(ctx, in, state, out)
               case DECODE_KEY => decodeKey(ctx, in, state)
               case DECODE_PARAMETERS => decodeParameters(ctx, in, state)
               case DECODE_VALUE => decodeValue(ctx, in, state)
            }
         }
      } catch {
         case e: Exception =>
            val (serverException, isClientError) = decodeCtx.createServerException(e, in)
            // If decode returns an exception, decode won't be called again so,
            // we need to fire the exception explicitly so that requests can
            // carry on being processed on same connection after a client error
            if (isClientError) {
               ctx.pipeline.fireExceptionCaught(serverException)
            } else {
               throw serverException
            }
         case t: Throwable => throw t
      }
   }

   private def decodeHeader(ctx: ChannelHandlerContext, buffer: ByteBuf, state: HotRodDecoderState, out: java.util.List[AnyRef]): AnyRef = {
      decodeCtx.header = new HotRodHeader
      val endOfOp = readHeader(buffer, decodeCtx.header)
      if (endOfOp.isEmpty) {
         // Something went wrong reading the header, so get more bytes.
         // It can happen with Hot Rod if the header is completely corrupted
         return null
      }
      val ch = ctx.channel
      decodeCtx.obtainCache(cacheManager)
      val cacheConfiguration = server.getCacheConfiguration(decodeCtx.header.cacheName)
      decodeCtx.defaultLifespanTime = cacheConfiguration.expiration().lifespan()
      decodeCtx.defaultMaxIdleTime = cacheConfiguration.expiration().maxIdle()
      if (endOfOp.get) {
         val message = decodeCtx.header.op match {
            case StatsRequest => writeResponse(ch, createStatsResponse)
            case _ => customDecodeHeader(ctx, buffer)
         }
         message match {
            case pr: PartialResponse => pr.buffer.map(out.add(_))
            case _ => null
         }
         null
      } else {
         checkpointTo(DECODE_KEY)
      }
   }

   private def decodeKey(ctx: ChannelHandlerContext, buffer: ByteBuf, state: HotRodDecoderState): AnyRef = {
      val ch = ctx.channel
      decodeCtx.header.op match {
         // Get, put and remove are the most typical operations, so they're first
         case GetRequest => writeResponse(ch, decodeCtx.get(readKey(buffer)._1))
         case PutRequest => handleModification(ch, buffer)
         case RemoveRequest => handleModification(ch, buffer)
         case GetWithVersionRequest => writeResponse(ch, decodeCtx.get(readKey(buffer)._1))
         case PutIfAbsentRequest | ReplaceRequest | ReplaceIfUnmodifiedRequest =>
            handleModification(ch, buffer)
         case _ => customDecodeKey(ctx, buffer)
      }
   }

   private def decodeParameters(ctx: ChannelHandlerContext, buffer: ByteBuf, state: HotRodDecoderState): AnyRef = {
      val ch = ctx.channel
      val endOfOp = readParameters(ch, buffer)
      checkpointTo(DECODE_VALUE)
      if (!endOfOp && decodeCtx.params.valueLength > 0) {
         // Create value holder only if there's more to read
         decodeCtx.rawValue = new Bytes(decodeCtx.params.valueLength)
         null
      } else if (decodeCtx.params.valueLength == 0) {
         decodeCtx.rawValue = Array.empty
         decodeValue(ctx, buffer, state)
      } else {
         decodeValue(ctx, buffer, state)
      }
   }

   private def decodeValue(ctx: ChannelHandlerContext, buffer: ByteBuf, state: HotRodDecoderState): AnyRef = {
      val ch = ctx.channel
      val ret = decodeCtx.header.op match {
         case PutRequest | PutIfAbsentRequest | ReplaceRequest | ReplaceIfUnmodifiedRequest =>
            buffer.readBytes(decodeCtx.rawValue)
            decodeCtx.header.op match {
               case PutRequest => decodeCtx.put
               case PutIfAbsentRequest => decodeCtx.putIfAbsent
               case ReplaceRequest => decodeCtx.replace
               case ReplaceIfUnmodifiedRequest => decodeCtx.replaceIfUnmodified
            }
         case RemoveRequest => decodeCtx.remove
         case _ => customDecodeValue(ctx, buffer)
      }
      writeResponse(ch, ret)
   }

   def readHeader(buffer: ByteBuf, header: HotRodHeader): Option[Boolean] = {
      try {
         val magic = buffer.readUnsignedByte
         if (magic != MAGIC_REQ) {
            if (!decodeCtx.isError) {
               throw new InvalidMagicIdException("Error reading magic byte or message id: " + magic)
            } else {
               trace("Error happened previously, ignoring %d byte until we find the magic number again", magic)
               return None // Keep trying to read until we find magic
            }
         }
      } catch {
         case e: Exception =>
            decodeCtx.isError = true
            throw e
      }

      val messageId = readUnsignedLong(buffer)
      val version = buffer.readUnsignedByte.toByte

      try {
         val decoder = version match {
            case VERSION_10 | VERSION_11 | VERSION_12 | VERSION_13 => Decoder10
            case VERSION_20 | VERSION_21 | VERSION_22 | VERSION_23 => Decoder2x
            case _ => throw new UnknownVersionException("Unknown version:" + version, version, messageId)
         }
         val endOfOp = decoder.readHeader(buffer, version, messageId, header)
         decodeCtx.decoder = decoder
         if (decodeCtx.isTrace) trace("Decoded header %s", header)
         decodeCtx.isError = false
         Some(endOfOp)
      } catch {
         case e: HotRodUnknownOperationException =>
            decodeCtx.isError = true
            throw e
         case e: Exception =>
            decodeCtx.isError = true
            throw new RequestParsingException("Unable to parse header", version, messageId, e)
      }
   }

   def readKey(b: ByteBuf): (Bytes, Boolean) = decodeCtx.decoder.readKey(decodeCtx.header, b)

   def readParameters(ch: Channel, b: ByteBuf): Boolean = {
      val (parameters, endOfOp) = decodeCtx.decoder.readParameters(decodeCtx.header, b)
      decodeCtx.params = parameters
      endOfOp
   }

   protected def customDecodeHeader(ctx: ChannelHandlerContext, buffer: ByteBuf): AnyRef =
      writeResponse(ctx.channel, decodeCtx.decoder.customReadHeader(decodeCtx.header,  buffer, decodeCtx.cache, server, ctx))

   protected def customDecodeKey(ctx: ChannelHandlerContext, buffer: ByteBuf): AnyRef =
      writeResponse(ctx.channel, decodeCtx.decoder.customReadKey(this, decodeCtx.header, buffer, decodeCtx.cache, server, ctx.channel))

   protected def customDecodeValue(ctx: ChannelHandlerContext, buffer: ByteBuf): AnyRef =
      writeResponse(ctx.channel, decodeCtx.decoder.customReadValue(this, decodeCtx.header, decodeCtx, buffer, decodeCtx.cache))

   def createStatsResponse: Response =
      decodeCtx.decoder.createStatsResponse(decodeCtx.header, decodeCtx.cache.getStats, transport)

   private def wrapSecurity(block: => Unit) = {
      if (secure) Security.doAs(subject, new PrivilegedExceptionAction[Unit] {
         override def run(): Unit = block
      })
      else block
   }

   def handleModification(ch: Channel, buf: ByteBuf): AnyRef = {
      val (k, endOfOp) = readKey(buf)
      decodeCtx.key = k
      if (endOfOp) {
         // If it's the end of the operation, it can only be a remove
         writeResponse(ch, decodeCtx.remove)
      } else {
         checkpointTo(DECODE_PARAMETERS)
      }
   }

   protected def writeResponse(ch: Channel, response: AnyRef): AnyRef = {
       if (response != null) {
         try {
            if (decodeCtx.isTrace) trace("Write response %s", response)
            response match {
               // We only expect Lists of ChannelBuffer instances, so don't worry about type erasure
               case l: Array[ByteBuf] =>
                  l.foreach(ch.write(_))
                  ch.flush
               case a: Bytes => ch.writeAndFlush(wrappedBuffer(a))
               case cs: CharSequence => ch.writeAndFlush(Unpooled.copiedBuffer(cs, CharsetUtil.UTF_8))
               case pr: PartialResponse => return pr
               case _ => ch.writeAndFlush(response)
            }
         } finally {
           resetParams
         }
      }
         null
   }

   private def resetParams() = {
      checkpointTo(DECODE_HEADER)
      // Reset parameters to avoid leaking previous params
      // into a request that has no params
      decodeCtx.resetParams()
      null
   }

   override def exceptionCaught(ctx: ChannelHandlerContext, cause: Throwable) {
      decodeCtx.exceptionCaught(ctx, cause)(resetParams())
   }

   def checkpointTo(state: HotRodDecoderState): AnyRef = {
      checkpoint(state)
      null // For netty's decoder that mandates a return
   }

   override def checkpoint() = {
     super.checkpoint
   }
}

class UnknownVersionException(reason: String, val version: Byte, val messageId: Long)
extends StreamCorruptedException(reason)

class HotRodUnknownOperationException(reason: String, val version: Byte, val messageId: Long)
extends UnknownOperationException(reason)

class InvalidMagicIdException(reason: String) extends StreamCorruptedException(reason)

class RequestParsingException(reason: String, val version: Byte, val messageId: Long, cause: Exception)
extends IOException(reason, cause) {
   def this(reason: String, version: Byte, messageId: Long) = this(reason, version, messageId, null)
}

class HotRodHeader {
   var op: Enumeration#Value = _
   var version: Byte = _
   var messageId: Long = _
   var cacheName: String = _
   var flag: Int = _
   var clientIntel: Short = _
   var topologyId: Int = _

   override def toString = {
      new StringBuilder().append("HotRodHeader").append("{")
      .append("op=").append(op)
      .append(", version=").append(version)
      .append(", messageId=").append(messageId)
      .append(", cacheName=").append(cacheName)
      .append(", flag=").append(flag)
      .append(", clientIntelligence=").append(clientIntel)
      .append(", topologyId=").append(topologyId)
      .append("}").toString
   }
}

class CacheNotFoundException(msg: String, override val version: Byte, override val messageId: Long)
extends RequestParsingException(msg, version, messageId)

class HotRodException(val response: ErrorResponse, cause: Throwable) extends Exception(cause)

class UnknownOperationException(reason: String) extends StreamCorruptedException(reason)

class PartialResponse(val buffer: Option[ByteBuf])

File: server/hotrod/src/main/java/org/infinispan/server/hotrod/HotRodDecoder.java
package org.infinispan.server.hotrod;

import java.util.List;
import java.util.function.Predicate;

import org.infinispan.commons.logging.LogFactory;
import org.infinispan.commons.util.Util;
import org.infinispan.manager.EmbeddedCacheManager;
import org.infinispan.server.core.logging.Log;
import org.infinispan.server.core.transport.ExtendedByteBufJava;
import org.infinispan.server.core.transport.NettyTransport;

import io.netty.buffer.ByteBuf;
import io.netty.buffer.ByteBufUtil;
import io.netty.channel.ChannelHandlerContext;
import io.netty.handler.codec.ByteToMessageDecoder;

/**
 * Decoder that will decode hotrod messages and then send a {@link CacheDecodeContext} down the pipeline.
 *
 * @author wburns
 * @since 9.0
 */
public class HotRodDecoder extends ByteToMessageDecoder {
   private final static Log log = LogFactory.getLog(HotRodDecoder.class, Log.class);

   private final EmbeddedCacheManager cacheManager;
   private final NettyTransport transport;
   private final Predicate<? super String> ignoreCache;
   private final HotRodServer server;

   CacheDecodeContext decodeCtx;
   Throwable previousException;

   private HotRodDecoderState state = HotRodDecoderState.DECODE_HEADER;

   private boolean resetRequested = true;

   public HotRodDecoder(EmbeddedCacheManager cacheManager, NettyTransport transport, HotRodServer server,
                        Predicate<? super String> ignoreCache) {
      this.cacheManager = cacheManager;
      this.transport = transport;
      this.ignoreCache = ignoreCache;
      this.server = server;

      this.decodeCtx = new CacheDecodeContext(server);
   }

   public NettyTransport getTransport() {
      return transport;
   }

   void resetNow() {
      decodeCtx = new CacheDecodeContext(server);
      decodeCtx.header = new HotRodHeader();
      state = HotRodDecoderState.DECODE_HEADER;
      resetRequested = false;
   }

   /**
    * Should be called when state is transferred.  This also marks the buffer read position so when we can't read bytes
    * it will be reset to here.
    *
    * @param newState new hotrod decoder state
    * @param buf      the byte buffer to mark
    */
   protected void state(HotRodDecoderState newState, ByteBuf buf) {
      buf.markReaderIndex();
      state = newState;
   }

   @Override
   protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
      try {
         if (CacheDecodeContext.isTrace) {
            log.tracef("Decode buffer %s using instance @%x", dumpHexByteBuf(in), System.identityHashCode(this));
         }

         if (resetRequested) {
            if (CacheDecodeContext.isTrace)
               log.tracef("Reset cached decoder data: %s", decodeCtx);
            resetNow();
         }

         // Mark the index to the beginning, just in case
         in.markReaderIndex();

         switch (state) {
            // These are all fall through cases which means they call to the one below if they needed additional
            // processing
            case DECODE_HEADER:
               if (!decodeHeader(in, out)) {
                  break;
               }
               state(HotRodDecoderState.DECODE_KEY, in);
            case DECODE_KEY:
               if (!decodeKey(in, out)) {
                  break;
               }
               state(HotRodDecoderState.DECODE_PARAMETERS, in);
            case DECODE_PARAMETERS:
               if (!decodeParameters(in, out)) {
                  break;
               }
               state(HotRodDecoderState.DECODE_VALUE, in);
            case DECODE_VALUE:
               if (!decodeValue(in, out)) {
                  break;
               }
               state(HotRodDecoderState.DECODE_HEADER, in);
               break;

            // These are terminal cases, meaning they only perform this operation and break
            case DECODE_HEADER_CUSTOM:
               readCustomHeader(in, out);
               break;
            case DECODE_KEY_CUSTOM:
               readCustomKey(in, out);
               break;
            case DECODE_VALUE_CUSTOM:
               readCustomValue(in, out);
               break;
         }
      } catch (Throwable t) {
         previousException = t;
         resetRequested = true;
         // Faster than throwing exception
         ctx.pipeline().fireExceptionCaught(new HotRodException(decodeCtx.createExceptionResponse(t), t.getMessage(), t));
      }
   }

   private static String dumpHexByteBuf(ByteBuf in) {
      int maxLength = 32;
      StringBuilder sb = new StringBuilder(maxLength * 2 + 20);
      sb.append('(').append(in.readableBytes()).append(')');
      int startIndex;
      if (in.readableBytes() < maxLength) {
         startIndex = in.readerIndex();
      } else {
         startIndex = in.writerIndex() - maxLength;
         sb.append("...");
      }
      for (int i = startIndex; i < in.writerIndex(); i++) {
         Util.addHexByte(sb, in.getByte(i));
      }
      return sb.toString();
   }

   boolean decodeHeader(ByteBuf in, List<Object> out) throws Exception {
      boolean shouldContinue = readHeader(in);
      // If there was nothing present it means we throw this decoding away and start fresh
      if (!shouldContinue) {
         return false;
      }
      HotRodHeader header = decodeCtx.header;
      // Check if this cache can be accessed or not
      if (ignoreCache.test(header.cacheName)) {
         throw new CacheUnavailableException();
      }
      decodeCtx.obtainCache(cacheManager);
      HotRodOperation op = header.op;
      switch (op.getDecoderRequirements()) {
         case HEADER_CUSTOM:
            state(HotRodDecoderState.DECODE_HEADER_CUSTOM, in);
            readCustomHeader(in, out);
            return false;
         case HEADER:
            // If all we needed was header, we have everything already!
            out.add(decodeCtx);
            resetRequested = true;
            return false;
         default:
            // Continue to key
            return true;
      }
   }

   /**
    * Reads the header and returns whether we should try to continue
    *
    * @param buffer the buffer to read the header from
    * @return whether or not we should continue
    * @throws Exception
    */
   boolean readHeader(ByteBuf buffer) throws Exception {
      VersionedDecoder decoder = decodeCtx.decoder;
      HotRodHeader header = decodeCtx.header;
      if (decoder == null) {
         if (buffer.readableBytes() < 1) {
            buffer.resetReaderIndex();
            return false;
         }
         short magic = buffer.readUnsignedByte();

         if (CacheDecodeContext.isTrace)
            log.tracef("Header magic: %d", magic);

         if (magic != Constants.MAGIC_REQ) {
            if (previousException == null) {
               dumpBuffer("Invalid magic id", buffer);
               throw new InvalidMagicIdException("Error reading magic byte or message id: " + magic);
            } else {
               if (CacheDecodeContext.isTrace) {
                  log.tracef("Error happened previously, ignoring %d byte until we find the magic number again", magic);
               }
               return false;
            }
         } else {
            previousException = null;
         }

         long messageId = ExtendedByteBufJava.readMaybeVLong(buffer);
         if (messageId == Long.MIN_VALUE) {
            return false;
         }

         if (CacheDecodeContext.isTrace) {
            log.tracef("Header message id: %d", messageId);
         }

         header.messageId = messageId;
         if (buffer.readableBytes() < 1) {
            buffer.resetReaderIndex();
            return false;
         }
         byte version = (byte) buffer.readUnsignedByte();
         header.version = version;

         if (CacheDecodeContext.isTrace) {
            log.tracef("Header version: %d", version);
         }

         decoder = HotRodVersion.getDecoder(version);
         if (decoder == null) {
            throw new UnknownVersionException("Unknown version:" + version, version, messageId);
         }
         decodeCtx.decoder = decoder;
         // This way we won't have to reread the decoder related material again
         buffer.markReaderIndex();
      }

      try {
         if (!decoder.readHeader(buffer, header.version, header.messageId, header)) {
            return false;
         }
         if (CacheDecodeContext.isTrace) {
            log.tracef("Decoded header %s", header);
         }
         return true;
      } catch (HotRodUnknownOperationException | SecurityException e) {
         throw e;
      } catch (Exception e) {
         throw new RequestParsingException("Unable to parse header", header.version, header.messageId, e);
      }
   }

   private static void dumpBuffer(String prefix, ByteBuf in) {
      if (log.isTraceEnabled()) {
         String dump = ByteBufUtil.hexDump(in).toUpperCase();
         log.tracef("%s error encountered, the buffer contains: %s", prefix, dump);
      }
   }

   private void readCustomHeader(ByteBuf in, List<Object> out) {
      decodeCtx.decoder.customReadHeader(decodeCtx.header, in, decodeCtx, out);
      // If out was written to, it means we read everything, else we have to reread again
      if (!out.isEmpty()) {
         resetRequested = true;
      }
   }

   boolean decodeKey(ByteBuf in, List<Object> out) {
      HotRodOperation op = decodeCtx.header.op;
      // If we want a single key read that - else we do try for custom read
      if (op.requiresKey()) {
         byte[] bytes = ExtendedByteBufJava.readMaybeRangedBytes(in);
         // If the bytes don't exist then we need to reread
         if (bytes != null) {
            if (CacheDecodeContext.isTrace) {
               log.tracef("Body key: %s", Util.printArray(bytes));
            }
            decodeCtx.key = bytes;
         } else {
            return false;
         }
      }
      switch (op.getDecoderRequirements()) {
         case KEY_CUSTOM:
            state(HotRodDecoderState.DECODE_KEY_CUSTOM, in);
            readCustomKey(in, out);
            return false;
         case KEY:
            out.add(decodeCtx);
            resetRequested = true;
            return false;
         default:
            return true;
      }
   }

   private void readCustomKey(ByteBuf in, List<Object> out) {
      decodeCtx.decoder.customReadKey(decodeCtx.header, in, decodeCtx, out);
      // If out was written to, it means we read everything, else we have to reread again
      if (!out.isEmpty()) {
         resetRequested = true;
      }
   }

   boolean decodeParameters(ByteBuf in, List<Object> out) {
      CacheDecodeContext.RequestParameters params = decodeCtx.decoder.readParameters(decodeCtx.header, in);
      if (params != null) {
         if (CacheDecodeContext.isTrace) {
            log.tracef("Body parameters: %s", params);
         }
         decodeCtx.params = params;
         if (decodeCtx.header.op.getDecoderRequirements() == DecoderRequirements.PARAMETERS) {
            out.add(decodeCtx);
            resetRequested = true;
            return false;
         }
         return true;
      } else {
         return false;
      }
   }

   boolean decodeValue(ByteBuf in, List<Object> out) {
      HotRodOperation op = decodeCtx.header.op;
      if (op.requireValue()) {
         int valueLength = decodeCtx.params.valueLength;
         if (in.readableBytes() < valueLength) {
            return false;
         }
         byte[] bytes = new byte[valueLength];
         in.readBytes(bytes);
         decodeCtx.operationDecodeContext = bytes;
         if (CacheDecodeContext.isTrace) {
            log.tracef("Body value: %s", Util.printArray(bytes));
         }
      }
      switch (op.getDecoderRequirements()) {
         case VALUE_CUSTOM:
            state(HotRodDecoderState.DECODE_VALUE_CUSTOM, in);
            readCustomValue(in, out);
            return false;
         case VALUE:
            out.add(decodeCtx);
            resetRequested = true;
            return false;
      }

      return true;
   }

   private void readCustomValue(ByteBuf in, List<Object> out) {
      decodeCtx.decoder.customReadValue(decodeCtx.header, in, decodeCtx, out);
      // If out was written to, it means we read everything, else we have to reread again
      if (!out.isEmpty()) {
         resetRequested = true;
      }
   }
}
